### DISCUSSION TITLE
Exploring CMD 0x0516 in firmware flashing mode

### DISCUSSION THREAD

**Discussion by amnemonic** [Tunas1337/UV-K5-Modded-Firmwares#23] [Ideas]
Hello, I want to start this thread to find out something about command 0x0516 which can be found in code responsible for flashing firmware update. This command length body is 104 bytes (0x68). Decompiled code which handles this function looks like that:

```cpp
  memcopy(0x20000FC4, uart_cmd_offset + 4, 104); //first 2 bytes are 0x16 0x05, next two are payload size so we skip 4 bytes in total and copy cmd payload to memory at address @0x20000FC4
  __disable_irq();
JUMPOUT(0x20000380);  //jump to code stored in RAM memory
  __enable_irq();
  return uart_cmd_0517rep(0); //reply to command 0x0516
```

does anybody have any idea about what that command does or how to get that code from ram memory?

**Comment by agocsdaniel** [comment]
0xfa34 - 0xfddc is mapped between 0x20000000 - 0x200003a8
(It's shifted with 1 byte for me, but that's fine I guess)
So:
0x01: FlashPreinit
0x8d: FlashInit
0x115: FlashSetAddr and something
0x225: FlashWrite
0x295: Init_ChipId_Pmu_CLK
0x381: FlashWriteNVR

What you are looking for is the last one
signature: FlashWriteNVR(u4 start_addr, u4 buffer, u4 length)

**Reply by amnemonic** [reply]
How is it _"mapped"_? Or is it copied?

**Reply by agocsdaniel** [reply]
Sorry, yes, it's copied.
By the entry routine. At 0xf1b0

**Reply by amnemonic** [reply]
I was looking at that code and I don't know if I get it disassembled correctly but for me it copies bytes from 0x0A04 to 0A034? Also, looking at addresses I think that boot code remaps flash at address range `0xF000 - 0xFFFF` to `0x0000 - 0x0FFF`
```
sub_1B0:
06 4C                        LDR     R4, =addr__0x0A04
01 25                        MOVS    R5, #1
06 4E                        LDR     R6, =addr__0x0A34
05 E0                        B       loc_1C4

loc_1B8:
E3 68                        LDR     R3, [R4,#0xC]
07 CC                        LDM     R4!, {R0-R2}
2B 43                        ORRS    R3, R5
0C 3C                        SUBS    R4, #0xC
98 47                        BLX     R3
10 34                        ADDS    R4, #0x10

loc_1C4:
B4 42                        CMP     R4, R6
F7 D3                        BCC     loc_1B8
FF F7 7E FF                  BL      j_main


04 0A 00 00  off_1CC         DCD addr__0x0A04
34 0A 00 00  off_1D0         DCD addr__0x0A34
```

**Reply by agocsdaniel** [reply]
The function you see is a wrapper for the copying process. Actually the copying happens at 0xf8f2
The for loop at f1b0 loops through the descriptor of the copy process. It has 3 passes. First, code bytes are copied. Then data bytes, then the rest is filled with zeros. Between fa04-fa34 you'll find 3x4 4-byte numbers. If you decompile the function you'll see that it iterates by 4 numbers. In every iteration it calls the function at the forth number (which is an address) with the parameters found in first, second, third in this order. Those parameters are the source start address, the destination start address and the length (in this order).

The block is very likely remapped, so treat the block as it would be at 0x0. or decompile it separately.
